### PR TITLE
[Semi-Auto] Add reduction spmd rule

### PR DESCRIPTION
**Pull request by pkuzyc**
### PR types
New features

### PR changes
Others

### Description
Pcard-70448
1. Modify the ExtractAttr common function to deal with cases about the miss match of python and cpp type (e.g. python bool -> cpp int)
2. Add reduction op's spmd rule for inferring distributed attributes. Implement the InferForward function for reduction op, i.e. infer output tensor's distributed attributes from input tensors'.

### CODE REVIEW

File: paddle/fluid/distributed/auto_parallel/spmd_rules/rules.h [hunk 1]
```diff
@@ -24,6 +25,7 @@ namespace auto_parallel {
 
 // matmul rule
 REGISTER_SPMD_RULE(matmul, MatmulSPMDRule);
```
Comment (JZ-LIANG) [inline paddle/fluid/distributed/auto_parallel/spmd_rules/rules.h:30]: register with op_name specifically

reduce_sum, sum, reduce_max, reduce_min
mean

File: paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+      ninputs,
+      1,
+      phi::errors::InvalidArgument("The size of InputSpec in reduction must "
+                                   "be equal to 1, but got [%d].",
+                                   ninputs));
+  VerifySpecs(input_specs, "reduction");
+
+  // step1: Build Einsum Notation
+  bool keep_dim = ExtractAttr<bool>("keep_dim", attrs);
+  // bool keep_dim = false;
```
Comment (JZ-LIANG) [inline paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc]: remove it

File: paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  std::vector<int64_t> reduce_dims =
+      ExtractAttr<std::vector<int64_t>>("dim", attrs);
```
Comment (JZ-LIANG) [inline paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc]: where the "dim" come from ? should it be "axis" ?

Phi API:
- op : max
  args : (Tensor x, IntArray axis={}, bool keepdim=false)
- op : mean
  args : (Tensor x, IntArray axis={}, bool keepdim=false)
- op : sum
  args : (Tensor x, IntArray axis={}, DataType dtype=DataType::UNDEFINED, bool keepdim=false)

File: paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+  // initialize output dist_attr's process_mesh, batch_dim and dynamic dims with
+  // input dist_attr.
+  TensorDistAttr output_dist_attr =
+      CopyTensorDistAttrForOutput(input_specs[0].dist_attr());
+  output_dist_attr.set_dims_mapping(output_dims_mapping);
+
+  std::vector<TensorDistAttr> new_input_dist_attrs;
+  std::vector<TensorDistAttr> output_dist_attrs;
+  output_dist_attrs.emplace_back(output_dist_attr);
+  // step2.3: update the input dist_attr if reshard is needed. When
```
Comment (JZ-LIANG) [inline paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc]: the replicate logic for reduce axis is wrong.

if the op is a "linearity"(e.g. sum, all, mean), and the reduce dim is shared, there is no need to reshard the reduce axis as replicated, and we just need mark this axis  of output tensor as "Partial" by the sharding mesh dim.

if the op is a "non-linearity"(e.g. variance, ), replicate logic  is need.

File: paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc (hunk 3 above)
Comment (JZ-LIANG) [inline paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc]: there should be another attribute for reduce, 

"reduce_type": sum, max, min, mean
"linearity": true/false

File: paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+      new_input_dims_mapping[i] = -1;
+    }
+  }
+  TensorDistAttr new_input_dist_attr(input_specs[0].dist_attr());
+  new_input_dist_attr.set_dims_mapping(new_input_dims_mapping);
+  new_input_dist_attrs.emplace_back(new_input_dist_attr);
+
+  // step2.4: handle partial
+  // Step2.4.1 Output Partial
+  std::vector<int64_t> partial_on_dims =
```
Comment (JZ-LIANG) [inline paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc:96]: should use this logic to infer the partial dim.

if a axis is missing in output tensor, and this axis is sharded in input tensor, 
this axis would be Partial on the dim in output tensor.

File: paddle/fluid/distributed/auto_parallel/spmd_rules/rules.h (hunk 1 above)
Comment (pkuzyc) [reply paddle/fluid/distributed/auto_parallel/spmd_rules/rules.h:30]: Done

File: paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc (hunk 2 above)
Comment (pkuzyc) [reply paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc]: Done

File: paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc (hunk 3 above)
Comment (pkuzyc) [reply paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc]: modified to "axis" now. "dim" is the attribute name in static mode.

File: paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc (hunk 4 above)
Comment (pkuzyc) [reply paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc]: Done

File: paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc (hunk 3 above)
Comment (pkuzyc) [reply paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc]: added a "linearity" attribute.

File: paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc (hunk 5 above)
Comment (pkuzyc) [reply paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc:96]: Done

File: paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+  TensorDistAttr output_dist_attr =
+      CopyTensorDistAttrForOutput(input_specs[0].dist_attr());
+  output_dist_attr.set_dims_mapping(output_dims_mapping);
+
+  std::vector<TensorDistAttr> output_dist_attrs;
+  output_dist_attrs.emplace_back(output_dist_attr);
+
+  // step2.4: handle partial
+  // Step2.4.1 Output Partial
+  // If the op is a linear op, i.e. `linearity` is true, the output's
```
Comment (JZ-LIANG) [inline paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc]: non-linear op requires its input to be non-partial. but could generate partial output.

Comment (pkuzyc) [reply paddle/fluid/distributed/auto_parallel/spmd_rules/reduction_spmd_rule.cc]: Done

**Review by JZ-LIANG — APPROVED** [review]
LGTM

**Pull request merged**